pkg/envoy/ads/stream.go

```diff
@@ -190,23 +193,62 @@ func respondToRequest(proxy *envoy.Proxy, discoveryRequest *xds_discovery.Discov
 	// than what we last sent. (Perhaps the control plane restarted.)
 	// In that case we want to make sure that we send new responses with
 	// VersionInfo incremented starting with the version which the proxy last had.
-	if ackVersion > proxy.GetLastSentVersion(typeURL) {
-		proxy.SetLastSentVersion(typeURL, ackVersion)
+	lastVersion = proxy.GetLastSentVersion(typeURL)
+	if requestVersion > lastVersion {
+		log.Debug().Msgf("Proxy SerialNumber=%s PodUID=%s: Higher version on request %s, req ver: %d - last ver: %d. Updating to match latest.",
+			proxy.GetCertificateSerialNumber(), proxy.GetPodUID(), typeURL.Short(), requestVersion, lastVersion)
+		proxy.SetLastSentVersion(typeURL, requestVersion)
+		return true
 	}
 
-	lastNonce := proxy.GetLastSentNonce(typeURL)
-	if lastNonce != "" && discoveryRequest.ResponseNonce == lastNonce {
-		log.Debug().Msgf("Proxy SerialNumber=%s PodUID=%s: Nothing changed for Envoy on Pod since Nonce=%s",
-			proxy.GetCertificateSerialNumber(), proxy.GetPodUID(), discoveryRequest.ResponseNonce)
+	// Compare Nonces
+	// As per protocol, we can ignore any request on the TypeURL stream that has not caught up with last sent nonce, if the
+	// nonce is non-empty.
+	// A newly connecting envoy might send a version and nonce from an old control plane, this is currently mitigated by the
+	// forced update @line 69
+	// TODO: Fix reconnection. Forced update causes other issues, we should stick to the protocol and wait for the first request
+	// and gracefully handle reconnecting envoys.
+	lastNonce = proxy.GetLastSentNonce(typeURL)
+	if requestNonce != lastNonce {
+		log.Debug().Msgf("Proxy SerialNumber=%s PodUID=%s: Ignoring request for %s non-latest nonce (request: %s, current: %s)",
+			proxy.GetCertificateSerialNumber(), proxy.GetPodUID(), typeURL.Short(), requestNonce, lastNonce)
 		return false
 	}
 
-	if discoveryRequest.ResponseNonce != "" {
-		log.Debug().Msgf("Proxy SerialNumber=%s PodUID=%s: Received discovery request with Nonce=%s matches=%t; proxy last Nonce=%s",
-			proxy.GetCertificateSerialNumber(), proxy.GetPodUID(), discoveryRequest.ResponseNonce, discoveryRequest.ResponseNonce == lastNonce, lastNonce)
+	// ----
+	// At this point, there is no error and nonces match, it is guaranteed an ACK with last version.
+	// What's left is to check if the resources listed are the same. If they are not, we must respond
+	// with the new reosurces requested
```

Comment:
```suggestion
	// with the new resources requested
```

typo and merge with L220?